### PR TITLE
Add application with LTI Provider

### PR DESCRIPTION
**Pull request by idegtiarov**
PR adds application with LTI Provider to the StarCellBio project

Documentation describing configuration on Provider and Consumer sides could be found in the document by the link: [LTI Configurations](https://docs.google.com/document/d/1vW0HEbjVvxdG6ezZM-mkuijiJMqemG-745IXSHtGdb0/edit?usp=sharing)

### CODE REVIEW

File: StarCellBio/settings.py [hunk 1]
```diff
@@ -112,7 +112,6 @@
 MIDDLEWARE_CLASSES = (
     'django.middleware.common.CommonMiddleware',
     'django.contrib.sessions.middleware.SessionMiddleware',
-    'django.middleware.csrf.CsrfViewMiddleware',
```
Comment (annagav) [inline StarCellBio/settings.py]: Why are we removing this?

Comment (idegtiarov) [reply StarCellBio/settings.py]: I've added this middleware in commit #31106d6 and found that it broke frontend tests that is why I removed it as it was in initial state.

File: lti_provider/lti_settings.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from django.conf import settings
+
+
+# Switch on debug mode for LTI requests
+DEBUG_LTI = getattr(settings, 'DEBUG_LTI', True)
+
+# Uncoment to switch off LTI ssl requirements (only for development mode)
```
Comment (auraz) [inline lti_provider/lti_settings.py]: UncomMent, typo.

File: lti_provider/models.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        self.save()
+        create_course_records(scb_user, course_id)
+
+    def login(self, request):
+        """
+        Login connected SCB user
+        """
+        if self.scb_user:
+            self.scb_user.backend = 'allauth.account.auth_backends.AuthenticationBackend'
+            logger.debug("Start User {} logining...".format(self.scb_user.username))
```
Comment (auraz) [inline lti_provider/models.py]: Typo, logging.

File: lti_provider/models.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            login(request, self.scb_user)
+            logger.debug("Check User is authenticated:: {}".format(request.user.is_authenticated()))
```
Comment (auraz) [inline lti_provider/models.py]: Rephrase plz.

File: lti_provider/unit_tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+from django.test import TestCase, RequestFactory
+
+from backend.models import Course, Assignment
+from lti_provider.models import Consumer, LTIUser
+from lti_provider.validator import RequestValidator
+from lti_provider.views import lti_launch
+
+
+def mock_tool_provider():
+    m = Mock()
```
Comment (auraz) [inline lti_provider/unit_tests/test_views.py:15]: single value names are restricted by pep8, i suppose.

File: lti_provider/validator.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        :param nonce: nonce from LTI request
+        :param request: LTI request
+        :return: boolean flag
+        """
+        logger.debug('Timestamp and nonce is validating...')
+        today = datetime.date.today()
+        consumer_expired_date = self.consumer.expiration_date
+        if consumer_expired_date and consumer_expired_date < today:
+            raise oauth1.OAuth1Error('Consumer Key is expired.')
+        return True
```
Comment (auraz) [inline lti_provider/validator.py:55]: Where is noonce validated here?

File: lti_provider/validator.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    def validate_client_key(self, client_key, request):
+        """
+        Validate client key ...
+
+        :param client_key: client key from LTI request
+        :param request: LTI request
+        :return: boolean flag
+        """
+        logger.debug('Client key is validating...')
```
Comment (auraz) [inline lti_provider/validator.py]: Please rephrase to be stateful.

File: lti_provider/views.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    app_description = 'Star Cell Bio LTI Application'
+    launch_view_name = 'lti:launch_experiment'
+    launch_url = request.build_absolute_uri(reverse(launch_view_name, args=[
+        'course_id',
+        'assignment_id',
+        'experiment'
+    ]))
+
+    # maybe you've got some extensions
+    # FIXME(idegtiarov) there is no extentions right now, if they don't appear remove example.
```
Comment (auraz) [inline lti_provider/views.py]: Please clarify this during pull request review, or change comment text.

File: lti_provider/views.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        tool_provider = DjangoToolProvider.from_django_request(request=request)
+        validator = RequestValidator()
+        ok = tool_provider.is_valid_request(validator)
+    except oauth1.OAuth1Error as err:
+        ok = False
+        logger.error('Error happened while LTI request: {}'.format(err.__str__()))
+    if settings.DEBUG_LTI:
+        logger.debug("LTI request is {}valid".format('' if ok else 'not '))
+    if not ok:
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Comment (auraz) [inline lti_provider/views.py]: Clean or fix TODO plz.

File: lti_provider/views.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        raise Http404('LTI request is not valid')
+
+    user_id = request_post.get('user_id')
+    if not user_id:
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Comment (auraz) [inline lti_provider/views.py]: Clean or fix TODO plz.

File: lti_provider/views.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    user, created = LTIUser.objects.get_or_create(user_id=user_id, consumer=consumer)
+
+    if not user.is_scb_user:
+        # NOTE(idegtiarov) connect user with the SCB user account
+        user.lti_to_scb_user(roles, course_id)
+        logger.debug('Check user was created {}'.format(user.is_scb_user))
+    user.login(request)
+    url = reverse('home')
+    if not course_id or not Course.objects.filter(code=course_id).exists():
+        # TODO(idegtiarov) add lti_error page if it is needed
```
Comment (auraz) [inline lti_provider/views.py]: Clean or fix TODO plz.

File: lti_provider/views.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        raise Http404('Course with the code {}, does not exist.'.format(course_id))
+    if assignment:
+        if not Assignment.objects.filter(assignmentID=assignment, courseID__code=course_id).exists():
+            # TODO(idegtiarov) add lti_error page if it is needed
```
Comment (auraz) [inline lti_provider/views.py]: Clean or fix TODO plz.

**Review by auraz — CHANGES_REQUESTED** [review]

File: lti_provider/views.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    if settings.DEBUG_LTI:
+        logger.debug(request.META)
+        logger.debug(request_post)
+
+    try:
+        tool_provider = DjangoToolProvider.from_django_request(request=request)
+        validator = RequestValidator()
+        ok = tool_provider.is_valid_request(validator)
+    except oauth1.OAuth1Error as err:
```
Comment (auraz) [inline lti_provider/views.py]: Can this error be raised here too DjangoToolProvider.from_django_request(request=request) ?

File: lti_provider/views.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    roles_from_request = request_post.get('roles', '').split(',')
+    roles = list({ROLES.get(role, 'student') for role in roles_from_request})
+    consumer = Consumer.objects.get(consumer_key=request_post['oauth_consumer_key'])
+
+    user, created = LTIUser.objects.get_or_create(user_id=user_id, consumer=consumer)
+
+    if not user.is_scb_user:
+        # NOTE(idegtiarov) connect user with the SCB user account
+        user.lti_to_scb_user(roles, course_id)
+        logger.debug('Check user was created {}'.format(user.is_scb_user))
```
Comment (auraz) [inline lti_provider/views.py]: please rephrase, what do you mean here?

File: lti_provider/lti_settings.py (hunk 2 above)
Comment (idegtiarov) [reply lti_provider/lti_settings.py]: done

File: lti_provider/models.py (hunk 3 above)
Comment (idegtiarov) [reply lti_provider/models.py]: fixed

File: lti_provider/validator.py (hunk 6 above)
Comment (idegtiarov) [reply lti_provider/validator.py:55]: We inherit oauth1.RequestValidator functions and timestamp is validated with `validate_timestamp_and_nonce` function. Nonce is not validating here.

File: lti_provider/validator.py (hunk 7 above)
Comment (idegtiarov) [reply lti_provider/validator.py]: done

File: lti_provider/views.py (hunk 8 above)
Comment (idegtiarov) [reply lti_provider/views.py]: We are not interested in any extension parameters, I remove this comment

File: lti_provider/views.py (hunk 9 above)
Comment (idegtiarov) [reply lti_provider/views.py]: Done

File: lti_provider/views.py (hunk 10 above)
Comment (idegtiarov) [reply lti_provider/views.py]: done

File: lti_provider/views.py (hunk 11 above)
Comment (idegtiarov) [reply lti_provider/views.py]: done

File: lti_provider/views.py (hunk 12 above)
Comment (idegtiarov) [reply lti_provider/views.py]: done

**Review by auraz — CHANGES_REQUESTED** [review]

File: lti_provider/validator.py (hunk 6 above)
Comment (auraz) [reply lti_provider/validator.py:55]: why it is passed and mentioned in name?

**Review by auraz — CHANGES_REQUESTED** [review]

File: lti_provider/views.py (hunk 13 above)
Comment (idegtiarov) [reply lti_provider/views.py]: Here can be raised two other errors InvalidLTIRequestError and ValueError. Add them in except checking block.

File: lti_provider/views.py (hunk 14 above)
Comment (idegtiarov) [reply lti_provider/views.py]: done

File: lti_provider/unit_tests/test_views.py (hunk 5 above)
Comment (idegtiarov) [reply lti_provider/unit_tests/test_views.py:15]: Not allowed to use only: l, O, and I variables.

File: lti_provider/validator.py (hunk 6 above)
Comment (idegtiarov) [reply lti_provider/validator.py:55]: Because in Oauth1 authentication this two params used and validated in pairs.

**Comment by annagav** [comment]
@idegtiarov, While deploying to our qa site there were at least a couple of dependencies missing in requirements.txt. Could you please update it?

**Comment by idegtiarov** [comment]
@annagav Yes sure I will check.

**Comment by idegtiarov** [comment]
@annagav I merge latest develop branch and run SCB in the new vagrant instance, requirements.txt covers all dependencies. If you know what packages should be added or updated please let me know.

**Comment by annagav** [comment]
Functionality works great! 👍

**Review by annagav — APPROVED** [review]

**Pull request merged**